### PR TITLE
Add initial support for external accounts

### PR DESCRIPTION
**Pull request by Xarthisius**
This PR providers a basic infrastructure for supporting auth integrations with 3rd party services using either OAuth flow, or API Key model (and DataONE which is a combination of both).

### Implementation details
The majority of the functionality is exposed via a single REST API endpoint: `GET /account?redirect=`. It returns an unsorted list of supported 3rd party services that user can authorize or is authorized with, as indicated by `state` field. Depending on `type` of auth provider (which can be `apikey`, `bearer` and `dataone`) there is a different flow that client UI should follow, as described below

#### OAuth (`type == 'bearer'`)
1. Authorization in OAuth flow, boils down to allowing user to follow a specially crafted href that jumps to OAuth IdP and then goes back to point of origin. If (`state == 'unauthorized'`), `url` field contains that url.
2. Deauthorization can be performed via `GET /account/:provider/revoke`. 

#### API Key (`type == 'apikey'`)
1. Authorization requires user to provide a key-value pair, where value is the API key, and the key is the `resource_server` that the former will be used with. A list of supported resource servers can be obtained by calling `GET /account/:provider/targets` e.g. `GET /account/zenodo/target`. After the key-value pair is provided by the user, it should be stored using `POST /account/:provider/key?resource_server=...&key=...`
2. Deauthorization is the same as in OAuth case, with the addition of `resource_server` as a required qarg.

#### DataONE (`type == 'dataone'`)
Authorization flow is similar to OAuth initially: you follow the url that accepts a redirect to a callback. The trick is: it doesn't return any of the usual information information, i.e. code nor token that can be used to obtain `access_token`. Instead DataONE sets a cookie on their page with _JSESSIONID_ that can be used to call a separate endpoint (`/portal/token`) to grab a JWT. I don't see a programmatic way of obtaining _JSESSIONID_ and that's why it has to happen on the client side... I came up with the following workflow:
1. User follows the `url` to authenticate with DataONE using ORCID authorization.
2. `url` sets the usual `redirect` to `/account/:provider/callback`, and during callback a bogus token is created:
   ```
   {
       "provider": "dataone...",
       "resource_server": <cn_netloc>,
       "token_type": "dataone-pre",
       "access_token": "",
   }
   ```
3. Whenever a user token of type _"dataone-pre"_ is detected in `GET /account`, the provider state is set to _"preauthorized"_
4. On the UI side, the settings page performs a `XMLHttpRequest` to obtain a JWT for each provider in state _"preauthorized"_ (see https://github.com/whole-tale/girder_wholetale/blob/ext_keys/web_client/views/ExtKeysView.js#L95-L125).
5. If a valid JWT is obtained in 4., it's stored as if it was an API key using `POST 'account/dataone.../key`. This updates both `token_type` and `access_token` in the existing token.

#### Notes
* `targets` keyword on providers returned by `GET /account` is only populated for authorized api key providers (check it's value during step 14. from *How to test?*) 

#### Requirements

* Grab ORCID client id/secret from `#core-dev` and add the following to your `.bashrc`:
   ```
   export ORCID_CLIENT_ID="APP-..."
   export ORCID_CLIENT_SECRET="6..."
   ```
   Make sure you're running latest `deploy-dev`. Run `source ~/.bashrc` before `make dev`.

### How to test?

1. Deploy locally, log in to Girder UI using Globus OAuth.
2. Navigate to _External Keys_ view (top right menu under your username) https://girder.local.wholetale.org/#ext_keys Should look like this (2 buttons + Zenodo + Dataverse):
extkeys

3. Open a terminal and run (keep it handy for verification that things work):
   ```
   docker exec -ti $(docker ps --filter=name=wt_girder -q) girder-shell
   ```
4. In terminal run:
   ```
   from girder.models.user import User
   User().findOne({'login': 'kowalikk'})['otherTokens']   # NOTE: change the login value
   ```
   At this point it should contain only Globus Transfer token.
5. On _External Keys_ page click on ORCID button.
6. After completed flow button should be visible in the 2nd row and 
   ```
   User().findOne({'login': 'kowalikk'})['otherTokens']
   ```
   should contain additional token.
7. Click on ORCID button in 2nd row.
8. It should return to the first row, token should be gone from user model.
9. Repeat 5. - 8. for DataONE
10. Click on + right of "Zenodo". Modal should open. Available choices for *Repository URL* should be: "zenodo.org" and "sanbox.zenodo.org".
11. Select "sandbox.zenodo.org", provide any value as API KEY, click "Connect"
12. Verify that "sandbox.zenodo.org" is displayed under Zenodo and it has red "trash can" button. Verify that user model has the appropriate token (check `resource_server` in particular).
13. Click on + right of "Zenodo" again. Modal should open. Available choices for *Repository URL* should be: "zenodo.org".
14. Repeat 11. - 12. using "zenodo.org"
15. Click on the red "trash can" button and confirm that selected resource disappear from UI and user model.

### CODE REVIEW

**Comment by bodom0015** [comment]
Testing with Swagger UI, I am getting a stack trace back when I execute `GET /account/{provider}/targets` with `provider="dataverse"`:
```json
{
  "message": "KeyError: KeyError('provider',)",
  "trace": [
    "<FrameSummary file /girder/girder/api/rest.py, line 630 in endpointDecorator>",
    "<FrameSummary file /girder/girder/api/rest.py, line 1210 in GET>",
    "<FrameSummary file /girder/girder/api/rest.py, line 967 in handleRoute>",
    "<FrameSummary file /girder/girder/api/access.py, line 63 in wrapped>",
    "<FrameSummary file /girder/girder/api/describe.py, line 709 in wrapped>",
    "<FrameSummary file /girder/plugins/wholetale/server/rest/account.py, line 303 in listTargetsForAccount>"
  ],
  "type": "internal"
}
```

I see that you've updated the branch, so let me retest

**Comment by bodom0015** [comment]
Also from Swagger UI, when attempting to `POST /account/{provider}/key`, I am not sure what to pass as the `resource_server` parameter:
```json
{
  "message": "Unsupported resource server \"dataverse.harvard.org\".",
  "type": "rest"
}
```

Any idea what I should even be passing here?

**Comment by Xarthisius** [comment]
It should be something like that:
```
curl -X POST \
  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  --header 'Girder-Token: ...' \
  https://girder.local.wholetale.org/api/v1/account/dataverse/key?resource_server=dataverse.harvard.org&key=blah&key_type=apikey
```

basically whichever value you got from `GET /account/{provider}/targets`.

**Comment by bodom0015** [comment]
After pulling again, `POST` appears to work correctly.

**Comment by bodom0015** [comment]
I may just be missing some things, but I'm still noticing the following:
* once connected, there does not appear to be a way to disconnect an account
* the mockups shows lines like `Authorized by k-davis` and `Authorized on zenodo.org by k-davis`, but this information is not currently being preserved
* `GET /account/dataone/targets` and/or `GET /account/dataonestage2/targets` return an error:
```json
{
  "message": "Unknown provider \"dataonestage2\".",
  "type": "rest"
}
```

**Comment by Xarthisius** [comment]
> * once connected, there does not appear to be an API method available to disconnect an account
That's True for DataONE. The only way to truly disconnect would be to 1) remove cookies from their site, 2) call `GET /account/dataone/revoke`. I guess we won't support that for now...

> * the mockups shows lines like `Authorized by k-davis` and `Authorized on zenodo.org by k-davis`, but this information is not currently being preserved or I am unable to find it

Fair. Since that would require a lot of additional work (it's provider specific), I'd suggest to skip that. @craig-willis thoughts?

> * While `dataverse` and `zenodo` appear to return correctly, `GET /account/dataone/targets` and/or `GET /account/dataonestage2/targets` return an error:

DataONE is not an API Key provider, but rather an (FAKE)OAuth one. It should only have one button (same as ORCID) per provider. There are multiple DataONE servers, but we treat them as separate entities.

**Comment by bodom0015** [comment]
Ahhhh... a secret hidden endpoint... There are several API calls missing from Swagger UI (e.g. `/revoke`, `/callback`, etc), which is likely why I was confused.

I could see it making sense for `/callback`, but is there a reason why `hide=True` is specified on `/revoke`?

**Comment by Xarthisius** [comment]
> Ahhhh... a secret hidden endpoint... There are several API calls missing from Swagger UI (e.g. `/revoke`, `/callback`, etc), which is likely why I was confused. As a result, I am only seeing three of the new endpoints listed there:
> 
> * `GET /account`
> * `POST /account/{provider}/key`
> * `GET /account/{provider}/targets`
> 
> It might make sense to hide `/callback`, but is there a reason why `hide=True` is specified on `/revoke`?

Apart from me copying and pasting it from the definition callback, probably not.

**Comment by bodom0015** [comment]
The property name `id` conflicts with many things in the UI (Ember-Data `id`, HTML element `id`, etc). I am unable read the value of the `id` field on the Ember side because it is being overwritten by something internal. Strangely, all other fields appear to be working. I may just be missing something obvious in the UI code, but the models themselves are fairly lightweight.

Girder typically uses `_id`, so it's difficult to say whether this is a common issue in the Ember UI or not. Would it be much work to possibly rename this field?

**Comment by Xarthisius** [comment]
> The property name `id` conflicts with many things in the UI (Ember-Data `id`, HTML element `id`, etc). I am unable read the value of the `id` field on the Ember side because it is being overwritten by something internal. Strangely, all other fields appear to be working. I may just be missing something obvious in the UI code, but the models themselves are fairly lightweight.
> 
> Girder typically uses `_id`, so it's difficult to say whether this is a common issue in the Ember UI or not. Would it be much work to possibly rename this field?

Sure, not a problem! Any preference towards the name?

**Comment by bodom0015** [comment]
I've thought a bit about it, but I have no strong preference and can't determine the most suitable alternative.. `key` is probably already too overloaded in the context of this feature, but `_id` may not be ideal because this is not a true Girder resource ID. We could use `slug`, but that's technically not correct either since it doesn't map to a URL path. `identifier` is already used elsewhere in the `dataset` model, so that has a different meaning assigned to it as well.

Any thoughts about what might be best here?

**Comment by Xarthisius** [comment]
How about: `id` -> `name`, `name` -> `fullName`?

**Comment by bodom0015** [comment]
Sure, that works for me :+1:

**Comment by bodom0015** [comment]
On `revoke`, I am getting the following:
```json
{"message": "Parameter \"redirect\" is required.", "type": "rest"}
```

Is this parameter required for removing an API key? seems strange, since I figured it would be ignored in that case.

**Comment by bodom0015** [comment]
The `DataONE Stage 2` provider is missing a `logo`.

And I noticed that `provider.targets` is always empty when performing a `GET /account`:
```
{
    "docs_href": "https://zenodo.org/account/settings/applications/tokens/new/",
    "fullName": "Zenodo",
    "logo": "abiglongbase64string",
    "name": "zenodo",
    "tags": [
      "data",
      "publish"
    ],
    "targets": [],
    "type": "apikey",
    "url": ""
  }
```

Was this intentional? Or is this field going away and should not be used?

**Comment by Xarthisius** [comment]
> EDIT: Should this `redirect` parameter always be a fully qualified URL? Apparently giving it just a URL suffix assumes that its a path relative to Girder. In the API key case, won't this endpoint always "redirect" back to the settings view (effectively a no-op)?

Yes, it has to be FQ url, otherwise it will behave as you described. I think it's a browser thing and the fact that we're hoping between subdomains, rather than something specific to girder.

> EDIT 2: I can't seem to pass a fully-qualified dashboard URL into the redirect parameter either:

Can you verify that's not an authentication issue by passing `token=<girderToken from cookie>` as an additional parameter?

> The DataONE provider is missing a logo.

Nope, that's most certainly an omission

> And I noticed that provider.targets is always empty when performing a GET /account:

You mean for DateONE (in this case it's like OAuth, there are no specific targets, only one-to-one mapping of the button and a provider), or Zenodo? For the latter, `targets` is not empty only if you provided an API key for any `resource_server`

**Comment by Xarthisius** [comment]
@bodom0015 I fixed some of the problems you've encountered by just removing that pointless `redirect` from `/revoke`. When I was implementing it, I was under the impression it's something that's needed in OAuth revocation flow. Turns out it's not...

**Comment by Xarthisius** [comment]
> * Why maintain a separate list of Dataverse installations for registration and publishing? Can't the list of targets just be based on the list of installations + extra hosts? Will we maintain a similar list for Zenodo for registration?

I assumed those sets may not be identical, i.e. we may read data from different places, than we write it to. The latter is answered by https://github.com/whole-tale/girder_wholetale/pull/365 (yes).

> * Following the OSF model (and what you have for Zenodo), the "Get Token" link would ideally point to the URL to create a token on the target, not the documentation (e.g., https://dataverse.harvard.edu/dataverseuser.xhtml?selectTab=apiTokenTab) instead of generic documentation. This would require templating the URL, since the host will change depending on selected target.

Agreed.

> * Am I correct that `GET /accounts/` returns a list of providers?  When I call `GET /account/{provider}/targets` for orcid or dataonestage2 I get a response  `Unknown provider`. Super minor, but I'd expect an empty list of targets, since the provider is known/valid.

I could return an empty list for non API Key, but I assumed that error is a nicer and stronger suggestion to API consumers that they shouldn't be doing that...

**Comment by bodom0015** [comment]
Running thru the test case, step 5 does not work and I receive the following error in the console:
```javascript
Uncaught TypeError: Cannot read property 'state' of undefined
    at child.clickGOauthButton (ExtKeysView.js:19)
    at executeBound (underscore.js:762)
    at HTMLDivElement.<anonymous> (underscore.js:775)
    at HTMLDivElement.<anonymous> (underscore.js:122)
    at HTMLDivElement.dispatch (jquery.js:5206)
    at HTMLDivElement.elemData.handle (jquery.js:5014)
```

I also noticed that my dataone `access_token` looks obviously incorrect:
```python
In [1]: from girder.models.user import User 
   ...: User().findOne({'login': 'lambert8'})['otherTokens']                                                                             
Out[1]: 
[{'access_token': 'Agd9zxO1Ebn24Jnjlj9rdoz46OOBvxGMpxGd28K4GG3ge8Q691cXCDYQYzn6ywKYP2gyejO27pyzEVfgEQ3p8TP86N',
  'expires_in': 172800,
  'refresh_token': 'AgPEWjK77pDWX769KqwPl0YQXeww4VVGz5EJdX0pq476j0neeMIxUrdNnbm4x6jGWNVWd5m96EKopdv3qP5Gz6rxWg1j7',
  'resource_server': 'transfer.api.globus.org',
  'scope': 'urn:globus:auth:scope:transfer.api.globus.org:all',
  'state': 'wGmaaeTgUGOxHJcWB4zYCyLpEGJyfBXvPn6DqkmAC4gQK5ObhsltQY0SOQeR1G3P.https://dashboard.local.wholetale.org?token={girderToken}',
  'token_type': 'Bearer'},
 {'access_token': '<!doctype html><html lang="en"><head><title>HTTP Status 404 – Not Found</title><style type="text/css">h1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} h2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} h3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} body {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} b {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} p {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;} a {color:black;} a.name {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 404 – Not Found</h1><hr class="line" /><p><b>Type</b> Status Report</p><p><b>Message</b> &#47;portal&#47;token&amp;target=https%3A&#47;&#47;girder.local.wholetale.org&#47;api&#47;v1&#47;account&#47;dataonestage2&#47;callback%3Fstate%3DYUNbi4bmccvOrBLyJZEg1etIQ0CIpg1jgcYJSWfRmGiZXYVTJvLETMKQojJbqE0i.https%253A&#47;&#47;girder.local.wholetale.org&#47;%2523ext_keys%26code%3Ddataone</p><p><b>Description</b> The origin server did not find a current representation for the target resource or is not willing to disclose that one exists.</p><hr class="line" /><h3>Apache Tomcat/8.5.39 (Ubuntu)</h3></body></html>',
  'provider': 'dataonestage2',
  'resource_server': 'cn-stage-2.test.dataone.org',
  'token_type': 'dataone'}]
```

Running `make clean && make dev` now to confirm there isn't just some weird state on my user

File: server/lib/dataone/fakeoauth.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from urllib.parse import urlparse, urlunparse, quote
+from d1_common.env import D1_ENV_DICT
+
+from girder.api.rest import getApiUrl
+from girder.plugins.oauth.providers.base import ProviderBase
+
+# Because, why would they keep up to date list of CNs and deployments?
+D1_ENV_DICT["stage2"] = dict(base_url="https://cn-stage-2.test.dataone.org/cn")
```
Comment (bodom0015) [inline server/lib/dataone/fakeoauth.py]: Original URL may be missing `/v2` at the end, as it is present here: https://github.com/whole-tale/dashboard/blob/master/app/components/dashboard/run/left-panel/component.js#L321

File: server/lib/dataone/fakeoauth.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    def getClientIdSetting(self):
+        return "fake_client_id"
+
+    def getClientSecretSetting(self):
+        return "fake_secret_id"
+
+    @classmethod
+    def getTokenUrl(cls):
+        return urlunparse(urlparse(cls.get_cn())._replace(path="/portal/token"))
```
Comment (bodom0015) [inline server/lib/dataone/fakeoauth.py]: If this `._replace` is replacing the full path suffix, then the full path should (correctly) be `https://cn-stage-2.test.dataone.org/portal/token`

Somehow this is still yielding a 404 for Girder :/

Comment (bodom0015) [reply server/lib/dataone/fakeoauth.py]: Scratch that, reading the error in the token shows a mangled URL suffix and hints at the problem: `&#47;portal&#47;token&amp;target=https%3A&#47;&#47;girder.local.wholetale.org&#47;api&#47;v1&#47;account&#47;dataonestage2&#47;callback%3Fstate%3D9t8QNP1PCO0HqtBv8uGOhyXGfQNtj1haqFuLr1Iit7Vb5XGoIck0DivBHJVAzLpx.https%253A&#47;&#47;girder.local.wholetale.org&#47;%2523ext_keys%26code%3Ddataone`

This long URL suffix should just be `/portal/token`, as there does not appear to be a `target` parameter on this page (NOTE: this is not a REST API, this is just Apache Tomcat)

Navigating to https://cn-stage-2.test.dataone.org/portal/token&target=https%3A%2F%2Fgirder.local.wholetale.org%2Fapi%2Fv1%2Faccount%2Fdataonestage2%2Fcallback%3Fstate%3D9t8QNP1PCO0HqtBv8uGOhyXGfQNtj1haqFuLr1Iit7Vb5XGoIck0DivBHJVAzLpx.https%253A%2F%2Fgirder.local.wholetale.org%2F%23ext_keys%26code%3Ddataone in the browser yields the same 404, but dropping the query string altogether yields a 200 response

Comment (Xarthisius) [reply server/lib/dataone/fakeoauth.py]: `/portal/token` endpoint should be called without any parameters. I'm confused what's going on. I'm gonna retest this branch to see if something's changed.

Comment (Xarthisius) [reply server/lib/dataone/fakeoauth.py]: @bodom0015 I tested it just now. Requests you should be seeing:

1. After clicking button for unauthorized state it goes to:

```
https://cn-stage-2.test.dataone.org/portal/oauth?action=start&target=https%3A//girder.local.wholetale.org/api/v1/account/dataonestage2/callback%3Fstate%3DkfL3zSIOrMPyWi0H.https%253A//girder.local.wholetale.org/%2523ext_keys%26code%3Ddataone
```

2. This goes back to `#ext_keys` page and in the dev console you can see a list of providers being outputed:

```
(4) [{…}, {…}, {…}, {…}]
0: {docs_href: "https://zenodo.org/account/settings/applications/tokens/new/", fullName: "Zenodo", logo: "/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAMCAgICAgMCAgIDAw…AQBAEAQBAEAQBAEAQBAEAQBAEAQBAEAQBAEAQBAEAQBAEB//Z", name: "zenodo", tags: Array(2), …}
1: {fullName: "ORCID", logo: "/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAkGBxAPDw8PEA8PDh…5ciIeUjyXJQitFXtE9VV3oj20EoulKIBCBJIAAAAAAAAAP//Z", name: "orcid", state: "unauthorized", tags: Array(1), …}
2:
fullName: "DataONE Stage 2 CN"
logo: "/9j/4AAQSkZJRgABAQAAAQABAAD/4gKgSUNDX1BST0ZJTEUAAQ"
name: "dataonestage2"
state: "preauthorized"
tags: ["publish"]
type: "dataone"
url: "https://cn-stage-2.test.dataone.org/portal/token"
__proto__: Object
3: {docs_href: "http://guides.dataverse.org/en/latest/api/auth.html", fullName: "Dataverse", logo: "/9j/4AAQSkZJRgABAQEASABIAAD/4QCARXhpZgAASUkqAAgAAA…qkXWwfFPrtoD2lTXkJhB6kKgyYLNtZIelA8KBgNgID/Ov/9k=", name: "dataverse", tags: Array(2), …}
length: 4
```
Note: state "preauthorized".
3. New state should trigger additional call to `/portal/token` for state "preauthorized":

```
https://cn-stage-2.test.dataone.org/portal/token
```
This gives 200 for me.

4. Token should be printed in the console :

```
Response from DataONE:
eyJhbGciOiJ...NEoTo866w
```

**Comment by craig-willis** [comment]
@bodom0015 I retested this with Kacper (via screenshare) in my dev environment and am not seeing the issues you're reporting.  We could try adding your key to my dev instance to see if it's possibly your account and/or you could test my credentials.

**Comment by bodom0015** [comment]
@craig-willis I'd be happy to sit down with you tomorrow to try and figure out what's happening here. As I mentioned, this still happens immediately after a `make clean && make dev`, and I am unsure of how else to reset my state.

If you're not seeing the same bugs, then I am definitely curious to see if the same would happen to my account on your system

**Pull request merged**